### ISSUE TITLE
Old primary was stuck in archive recovery state

### ISSUE THREAD

**Issue by yy8848** [zalando/patroni#2242]
**Describe the bug**
The cluster was setup by [Crunchy Data PGO](https://github.com/CrunchyData/postgres-operator).  Due to network partition, node running primary was isolated. Primary lost connection to DCS (K8s API server), repo and replicas. After that node was shutdown. Some time later network partition was fixed and old primary was restarted. But it was stuck in archive recovery state.

**Environment**
patronictl version 2.1.1

**Observation**

1. After network partition happened, primary received immediate shutdown request and was restarted. It entered into standby mode. But due to no connection to repo, the database recovered from local wal logs. 

```sql
2022-03-04 05:13:34.720 UTC [100] LOG:  received immediate shutdown request
...
2022-03-04 05:13:39.933 UTC [30234] LOG:  starting PostgreSQL 13.5 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-4), 64-bit
...
2022-03-04 05:13:49.984 UTC [30240] LOG:  entering standby mode
2022-03-04 05:13:49.984 UTC [30240] LOG:  database system was not properly shut down; automatic recovery in progress
2022-03-04 05:13:49.987 UTC [30240] LOG:  redo starts at 0/5014228
2022-03-04 05:13:49.988 UTC [30240] LOG:  invalid record length at 0/501E8B8: wanted 24, got 0
2022-03-04 05:13:49.990 UTC [30240] LOG:  consistent recovery state reached at 0/501E8B8
2022-03-04 05:13:49.991 UTC [30234] LOG:  database system is ready to accept read only connections
```
The mode is sync replication. But unfortunately the last wal log entry is a standby snapshot log.  So MinRecoveryPoint was updated in pg_control, which caused divergence later.
```sql
rmgr: Standby     len (rec/tot):     50/    50, tx:          0, lsn: 0/0501E880, prev 0/0501E850, desc: RUNNING_XACTS nextXid 672 latestCompletedXid 671 oldestRunningXid 672
pg_waldump: fatal: error in WAL record at 0/501E880: invalid record length at 0/501E8B8: wanted 24, got 0
```
In async mode, it leads to divergence in higher probability.

2. Node was shutdown. Old primary database received smart shutdown request. 
```sql
2022-03-04 05:17:25.377 UTC [30234] LOG:  received smart shutdown request
```
But the time was very short and database state was still in archive recovery.
```bash
Database cluster state:               in archive recovery
```
3. After some time, network partition issue was fixed. And primary was restarted. Since the old primary could connect to repo host, the database started to recover from archive. But it found that MinRecoveryPoint is not contained in wal logs in repo. So startup failed.

```sql
2022-03-04 05:25:05.290 UTC [85] LOG:  starting PostgreSQL 13.5 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-4), 64-bit
...
2022-03-04 05:25:05.302 UTC [88] LOG:  database system was interrupted while in recovery at log time 2022-03-04 05:05:16 ...
2022-03-04 05:25:05.541 UTC [88] LOG:  restored log file "00000002.history" from archive
2022-03-04 05:25:05.780 UTC [88] LOG:  entering standby mode
...
2022-03-04 05:25:06.310 UTC [102] FATAL:  the database system is starting up
2022-03-04 05:25:06.517 UTC [88] FATAL:  requested timeline 2 does not contain minimum recovery point 0/501E8B8 on timeline 1
...
2022-03-04 05:25:06.518 UTC [85] LOG:  startup process (PID 88) exited with exit code 1
2022-03-04 05:25:06.518 UTC [85] LOG:  aborting startup due to startup process failure
2022-03-04 05:25:06.520 UTC [85] LOG:  database system is shut down
```
4. In the log, we found Patroni failed to [get timeline and lsn](https://github.com/zalando/patroni/blob/075918d447cbe87c09eef76b875225134abfc382/patroni/postgresql/rewind.py#L111) since database was not running but the state is in recovery archive. So pg_rewind was not called.

```sql
2022-03-04 05:25:05,071 INFO: pg_controldata:
...
  Database cluster state: in archive recovery
...
  Minimum recovery ending location: 0/501E8B8
...
2022-03-04 05:25:05,080 INFO: Local timeline=None lsn=None
```
5. For each iteration in the loop, Patroni tried to restart the instance. But due to divergence, archive recovery always failed and startup was aborted. The member was stuck in such loop.

**Expected behavior**
Old primary can follow new primary after restart.

**Suggestion**
The issue was caused by non-proper shutdown. We found there is code to [ensure clean shutdown](https://github.com/zalando/patroni/blob/075918d447cbe87c09eef76b875225134abfc382/patroni/ha.py#L328). So pg_rewind can handle divergence afterwards. But ```in archive recovery``` is not considered. Just wondering whether it can be changed to also handle such condition.

**Comment by CyberDem0n** [comment]
Thank you @yy8848  for a nice analysis of the problem.

It seems to be a common issue that we also periodically see in our infrastructure.

If I think about how it should be solved, then probably we also need to consider including `in archive recovery` to the https://github.com/zalando/patroni/blob/075918d447cbe87c09eef76b875225134abfc382/patroni/postgresql/rewind.py#L115. Since the `Minimum recovery ending location` is updated while `in archive recovery` we can rely on it to figure out that the rewind is required. After that, if rewind is required, we need to do a clean shutdown and execute pg_rewind.

**Comment by yy8848** [comment]
I agree. That is another option. But in my mind, Patroni can also figure out the divergence in another round after the database is shutdown cleanly. My understanding is that MinRecoveryPoint will not change in single user mode（just hypothesis not validated). And the code is easier to follow. Seems that pg_rewind just [ensures clean shutdown](https://github.com/postgres/postgres/blob/a33e17f210547226ada52d2b8af851c3553bb4fa/src/bin/pg_rewind/pg_rewind.c#L316) before checking MinRecoveryPoint.

**Comment by yy8848** [comment]
I did not mention other phenomenon observed. If the database was shutdown normally (signal was handled correctly), the state would be changed to ```shut down in recovery```. Then the database could be started again without entering into the error loop. In the log, we saw that pg_rewind just re-inited the database. 
And even in the error loop, if database received TERM signal, state would also be changed into ```shut down in recovery``` (we just found that when hibernating and waking up K8s cluster by [Gardener](https://github.com/gardener/gardener)). The error loop was broken and database could also be restarted.
My understanding is that Patroni will execute branch https://github.com/zalando/patroni/blob/075918d447cbe87c09eef76b875225134abfc382/patroni/postgresql/rewind.py#L115 if the database is in correct shutdown state.
